README.md
# uuid62

[![npm version](https://badge.fury.io/js/uuid62.svg)](http://badge.fury.io/js/uuid62)
[![CircleCI](https://circleci.com/gh/shanehughes3/uuid62.svg?style=shield)](https://circleci.com/gh/shanehughes3/uuid62)
[![Dependency Status](https://david-dm.org/shanehughes3/uuid62.svg)](https://david-dm.org/shanehughes3/uuid62)
[![Known Vulnerabilities](https://snyk.io/test/github/shanehughes3/uuid62/badge.svg?targetFile=package.json)](https://snyk.io/test/github/shanehughes3/uuid62?targetFile=package.json)

Base-62 UUID generator

## Overview

`uuid62` makes it easy to generate short base-62 (or any other base) UUIDs. ID
generation is done by [uuid](https://github.com/kelektiv/node-uuid)
which follows [RFC4122](http://www.ietf.org/rfc/rfc4122.txt). The encoded UUIDs
are alphanumeric [0-9a-zA-Z] and always have a length of 22 chars.

`uuid62` began as a fork of
[dmarcelino](https://github.com/dmarcelino/uuid-base62)'s
`uuid-base62`. It provides updated dependencies and more features above that
package. `uuid62` is designed to be a drop-in replacement for `uuid-base62`,
however, so migration should be trivial.


## Installation
```shell
npm i uuid62
```

## Usage
```javascript
const uuid62 = require('uuid62');

const uuid = uuid62.v4();
// -> 2qY9COoAhfMrsH7mCyh86T

// decode a base-62 uuid
const originalUUID = uuid62.decode(uuid);
// -> 9af099b2-6244-4fc1-b72b-1d69a24481b7

// base-62 encode an existing traditional uuid
const encoded = uuid62.encode('8fc60e7c-3b3c-48e9-a6a7-a5fe4f1fbc31');
// -> 2fNwVYePN8WqqDFvVf7XMN
```

## API

`uuid62` is essentially a wrapper around [uuid](https://www.npmjs.com/package/uuid).
All optional parameters specified in `uuid`'s API can also be provided here.
Note that any buffer that would be returned by `uuid` will be converted into a
base-62 string representation by this library.

### v1 (timestamp-based)

`uuid62.v1()`

Returns a string. See `uuid` for optional parameters to specify timestamp and
node id.

### v4 (random)

`uuid62.v4()`

Returns a string.

### v5 (namespace)

`uuid62.v5(<name>, <namespace>)`

Returns a string. Requires:
- `name` - a string or array
- `namespace` - a string or buffer representing a uuid. String representation
  may be conventional or base-62. Two pre-defined namespaces are exposed at
  `uuid62.DNS` and `uuid62.URL`.

Examples:
```javascript
let id = uuid62.v5('https://google.com', uuid62.URL);

id = uuid62.v5('google.com' uuid62.DNS);

const myNS = uuid62.v4();
id = uuid62.v5('foobar', myNS);
```

### Encoding/decoding

`uuid62.encode(<id>, <encoding>)`

Returns a string representing a base-62 id. Takes:
- `id` - a string or array representing a conventional uuid
- `encoding` - an optional string specifying the encoding of the input id.
  Defaults to hex.

If the input id is of the incorrect length, the output will be padded with `0`
on the left or trimmed from the left.

`uuid62.decode(<id>, <encoding>)`

Returns a string representing a conventional uuid, including dashes. Takes:
- `id` - a string representing a base-62 uuid
- `encoding` - an optional string specifying the encoding of the output id.
  Defaults to hex.

If the input id is of the incorrect length, it will be padded with `0` on the
left or trimmed from the left before decoding.


### Other bases

uuid62 can support other bases by assigning a new baseX charset to `customBase`:
```javascript
uuid62.customBase = new uuid62.baseX("0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-_");
const uuid = uuid62.v4();
// -> 31LoSI_BVeQpXtwu_-GEbL
```

## License
MIT


CHANGELOG.md
# Changelog

## 1.0.2
- update dependencies

## 1.0.1
- Update dependencies
- Declare `Buffer` in module scope

## 1.0.0

- Implemented v5 (namespace) UUID creation
- Added tests for existing functionality
- Removed deprecated calls to parent `uuid` library methods
- Added more thorough API description in readme


test/uuid62.test.js
const chai = require('chai');
const should = chai.should();
const expect = chai.expect;
const uuid62 = require('../uuid62');
const { v5 } = require('uuid');

describe('uuid62', () => {
	describe('v4', () => {
		it('should generate an id without any params', () => {
			const id = uuid62.v4();
			id.should.not.be.null;
			id.should.be.a('String');
			id.should.match(/^[0-9A-Za-z]{22}$/);
		});

		it('should convert base-62 id back to uuid format', function () {
			const id = uuid62.v4();
			id.should.not.be.null;
			const uuid = uuid62.decode(id);
			uuid.should.not.be.null;
			uuid.should.be.a('String');
			uuid.length.should.equal(36);

			const parts = uuid.split('-');
			parts.length.should.equal(5);
			parts.forEach((group) => {
				const num = parseInt(group, 16);
				num.should.be.at.least(0);
			});
		});

		it('should generate same encoded uuid from different formats', () => {
			const uuidA = 'de305d54-75b4-431b-adb2-eb6b9e546014';
			const uuidB = new Buffer([0xde, 0x30, 0x5d, 0x54, 0x75, 0xb4, 0x43, 0x1b, 0xad, 0xb2, 0xeb, 0x6b, 0x9e, 0x54, 0x60, 0x14]);

			const idA = uuid62.encode(uuidA);
			const idB = uuid62.encode(uuidB);

			idA.should.equal(idB);
		});

		it('should match decoded UUID to original UUID', () => {
			const uuid = 'de305d54-75b4-431b-adb2-eb6b9e546014';

			const id = uuid62.encode(uuid);
			const output = uuid62.decode(id);

			output.should.equal(uuid);
		});
	});

	describe('v5', () => {
		it('should throw a TypeError when called without parameters', () => {
			expect(() => {
				const uuid = uuid62.v5();
			}).to.throw(TypeError);
		});

		it('should throw a TypeError when called without a namespace value', () => {
			expect(() => {
				const uuid = uuid62.v5('shanehughes.io');
			}).to.throw(TypeError);
		});

		it('should throw a TypeError when called with a non-uuid namespace value', () => {
			expect(() => {
				const uuid = uuid62.v5('shanehughes.io', 'DNS')
			}).to.throw(TypeError);
		});

		it('should expose pre-defined v5 namespaces', () => {
			uuid62.URL.should.equal('3h8Pgc0Wb7WH6HsyG77m40');
			uuid62.decode(uuid62.URL).should.equal(v5.URL);
			uuid62.DNS.should.equal('3h8PgalTIPNcNhUD4ZbIKY');
			uuid62.decode(uuid62.DNS).should.equal(v5.DNS);
		});

		it('should generate an id with a URL namespace', () => {
			uuid62.v5('https://shanehughes.io', uuid62.URL)
				.should.match(/^[0-9A-Za-z]{22}$/);
		});

		it('should generate an id with a DNS namespace', () => {
			uuid62.v5('shanehughes.io', uuid62.DNS)
				.should.match(/^[0-9A-Za-z]{22}$/);
		});

		it('should generate an id with a custom base-62 namespace', () => {
			uuid62.v5('custom-name', uuid62.v4())
				.should.match(/^[0-9A-Za-z]{22}$/);
		});

		it('should generate an id with a custom conventional namespace', () => {
			uuid62.v5('custom-name', '06ad547f-fe02-477b-9473-f7977e4d5e17')
				.should.match(/^[0-9A-Za-z]{22}$/);
		});
	});

	describe('v1', () => {
		it('should generate a unique id without any params', () => {
			const id = uuid62.v1();
			id.should.not.be.null;
			id.should.be.a('String');
			id.length.should.equal(22);
		});
	});

	describe('encode/decode', () => {
		const fixtures = {
			'0000000000000000000000': '00000000-0000-0000-0000-000000000000',
			'0cBaidlJ84Ggc5JA7IYCgv': '06ad547f-fe02-477b-9473-f7977e4d5e17',
			'4vqyd6OoARXqj9nRUNhtLQ': '941532a0-6be1-443a-a9d5-d57bdf180a52',
			'5FY8KwTsQaUJ2KzHJGetfE': 'ba86b8f0-6fdf-4944-87a0-8a491a19490e',
			'7N42dgm5tFLK9N8MT7fHC7': 'ffffffff-ffff-ffff-ffff-ffffffffffff'
		};
		let output;

		for (const key in fixtures) {
			it(`should properly encode ${fixtures[key]}`, () => {
				output = uuid62.encode(fixtures[key]);
				output.should.equal(key);
			});

			it(`should properly decode ${key}`, () => {
				output = uuid62.decode(key);
				output.should.equal(fixtures[key]);
			});
		}

		it('should handle uppercase characters in conventional uuids', () => {
			uuid62.encode('06AD547F-FE02-477B-9473-F7977E4D5E17')
				.should.equal('0cBaidlJ84Ggc5JA7IYCgv');
		});
	});

	describe('custom-base encoding', () => {
		it('should generate a unique id without any params in base 64', () => {
			uuid62.customBase = new uuid62.baseX("0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-_");
			const id = uuid62.v4();
			id.should.not.be.null;
			id.should.be.a('String');
			id.length.should.equal(22);
		});

		it('should encode and decode a uuid in base 64', () => {
			uuid62.customBase = new uuid62.baseX("0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-_");
			const uuid = '72be7291-fbf6-400f-87c4-455e23d01cd5';

			const id = uuid62.encode(uuid);
			id.should.not.be.null;
			id.should.equal('1OLDah-_p03Uv4hlUzQ1Pl');

			const output = uuid62.decode(id);
			output.should.not.be.null;
			output.should.equal(uuid);
		});
	});
});


test/uuid-base62.test.js
var assert = require('assert');
var uuidBase62 = require('../uuid-base62');

describe('uuid-base62', function () {
	describe('v4', function () {
		it('should generate a unique id without any params', function () {
			var res = uuidBase62.v4();
			assert(res);
			assert.equal(typeof res, 'string');
			assert.equal(res.length, 22);
		});

		it('should convert id back to uuid format', function () {
			var uuid = uuidBase62.v4();
			assert(uuid);

			var res = uuidBase62.decode(uuid);

			assert.equal(typeof res, 'string');
			assert.equal(res.length, 36);

			var parts = res.split('-');
			assert.equal(parts.length, 5);
			parts.forEach(function (group) {
				var num = parseInt(group, 16);
				assert(num > 0);
			});
		});

		it('should generate same encoded uuid from different formats', function () {
			var uuid = 'de305d54-75b4-431b-adb2-eb6b9e546014';
			var uuidB = new Buffer([0xde, 0x30, 0x5d, 0x54, 0x75, 0xb4, 0x43, 0x1b, 0xad, 0xb2, 0xeb, 0x6b, 0x9e, 0x54, 0x60, 0x14]);

			var res = uuidBase62.encode(uuid);
			var resB = uuidBase62.encode(uuidB);

			assert.equal(resB, res);
		});

		it('decoded UUID should match original UUID', function () {
			var uuid = 'de305d54-75b4-431b-adb2-eb6b9e546014';

			var uuidB62 = uuidBase62.encode(uuid);
			var res = uuidBase62.decode(uuidB62);

			assert.equal(res, uuid);
		});
	});

	describe('v1', function () {
		it('should generate a unique id without any params', function () {
			var res = uuidBase62.v1();
			assert(res);
			assert.equal(typeof res, 'string');
			assert.equal(res.length, 22);
		});
	});

	describe('encode / decode', function () {
		var fixtures = {
			'0000000000000000000000': '00000000-0000-0000-0000-000000000000',
			'0cBaidlJ84Ggc5JA7IYCgv': '06ad547f-fe02-477b-9473-f7977e4d5e17',
			'4vqyd6OoARXqj9nRUNhtLQ': '941532a0-6be1-443a-a9d5-d57bdf180a52',
			'5FY8KwTsQaUJ2KzHJGetfE': 'ba86b8f0-6fdf-4944-87a0-8a491a19490e',
			'7N42dgm5tFLK9N8MT7fHC7': 'ffffffff-ffff-ffff-ffff-ffffffffffff'
		};

		Object.keys(fixtures).forEach(function(key){
			it('should properly encode ' + fixtures[key], function () {
				assert.equal(uuidBase62.encode(fixtures[key]), key);
			});

			it('should properly decode ' + key, function () {
				assert.equal(uuidBase62.decode(key), fixtures[key]);
			});
		});
	});

	describe('other bases', function () {
		it('should generate a unique id without any params in base64', function () {
			uuidBase62.customBase = new uuidBase62.baseX("0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-_");
			var res = uuidBase62.v4();
			assert(res);
			assert.equal(typeof res, 'string');
			assert.equal(res.length, 22);
		});

		it('should encode and decode a uuid in Base64', function () {
			uuidBase62.customBase = new uuidBase62.baseX("0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-_");
			var uuid = '72be7291-fbf6-400f-87c4-455e23d01cd5';

			var uuidB64 = uuidBase62.encode(uuid);
			assert.equal(uuidB64, '1OLDah-_p03Uv4hlUzQ1Pl');

			var res = uuidBase62.decode(uuidB64);
			assert.equal(res, uuid);
		});
	});
});


uuid-base62.js
'use strict';

// dependencies
var uuid = require('node-uuid');
var baseX = require('base-x');
var base62 = baseX('0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ')

// expose node-uuid and baseX for convenience
module.exports.uuid = uuid;
module.exports.baseX = baseX;
module.exports.customBase = base62;
module.exports.length = 22;
module.exports.uuidLength = 32;

/**
* v4
*/
module.exports.v4 = function v4() {
	var self = this;
	var args = Array.prototype.slice.call(arguments);

	if (!args[1]) {
		// make sure we use a buffer to avoid getting an uuid with dashes
		args[1] = new Buffer(16);
	}

	var id = uuid.v4.apply(self, args);
	return self.encode(id);
};


/**
* v1
*/
module.exports.v1 = function v1() {
	var self = this;
	var args = Array.prototype.slice.call(arguments);

	if (!args[1]) {
		// make sure we use a buffer to avoid getting an uuid with dashes
		args[1] = new Buffer(16);
	}

	var id = uuid.v1.apply(self, args);
	return self.encode(id);
};


/**
* encode
*/
module.exports.encode = function encode(input, encoding) {
	encoding = encoding || 'hex';

	if (typeof input === 'string') {
		// remove the dashes to save some space
		input = new Buffer(input.replace(/-/g, ''), encoding);
	}
	return ensureLength(this.customBase.encode(input), module.exports.length);
};

/**
* decode
*/
module.exports.decode = function decode(b62Str, encoding) {
	encoding = encoding || 'hex';
	var res = ensureLength(new Buffer(this.customBase.decode(b62Str)).toString(encoding), module.exports.uuidLength);

	// re-add the dashes so the result looks like an uuid
	var resArray = res.split('');
	[8, 13, 18, 23].forEach(function(idx){
		resArray.splice(idx, 0, '-');
	});
	res = resArray.join('');

	return res;
};

/**
* ensureLength
*
* @api private
*/
function ensureLength(str, maxLen){
	str = str + "";
	if(str.length < maxLen){
		return padLeft(str, maxLen);
	}
	else if (str.length > maxLen){
		return trimLeft(str, maxLen);
	}
	return str;
};


/**
* padLeft
*
* @api private
*/
function padLeft(str, padding){
	str = str + "";
	var pad = "";
	for(var i=str.length; i<padding; i++){
		pad += '0';
	}
	return pad + str;
};

/**
* trimLeft
*
* @api private
*/
function trimLeft(str, maxLen){
	str = str + "";
	var trim = 0;
	while(str[trim] === '0' && (str.length - trim) > maxLen){
		trim++;
	}
	return str.slice(trim);
};


uuid62.js
'use strict';

const uuid = require('uuid');
const uuidv1 = uuid.v1;
const uuidv4 = uuid.v4;
const uuidv5 = uuid.v5;
const baseX = require('base-x');
const base62 = baseX('0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ');
const OUTPUT_LENGTH = 22;
const UUID_LENGTH = 32;
let Buffer = (global) ? global.Buffer : undefined;

if (typeof Buffer === 'undefined') {
	// import buffer module for use in browser - browserify and perhaps webpack
	// should do this automatically, but otherwise we'll pull it in here
	Buffer = require('buffer').Buffer;
}

class UUID62 {
	constructor() {
		this.base = base62;
		// expose underlying methods for convenience (matches uuid-base62)
		this.uuid = uuid;
		this.baseX = baseX;
		// pre-defined v5/v3 namespaces
		this.URL = '3h8Pgc0Wb7WH6HsyG77m40';
		this.DNS = '3h8PgalTIPNcNhUD4ZbIKY';
	}

	get customBase() {
		return this.base;
	}

	set customBase(base) {
		this.base = base;
	}

	get length() {
		return OUTPUT_LENGTH;
	}

	get uuidLength() {
		return UUID_LENGTH;
	}

	v4() {
		const args = Array.prototype.slice.call(arguments);
		if (!args[1]) {
			// make sure we use a buffer to avoid getting a uuid with dashes
			args[1] = new Buffer(16);
		}
		const id = uuidv4.apply(this, args);
		return this.encode(id);
	}

	v5() {
		const args = Array.prototype.slice.call(arguments);
		if (typeof args[1] === 'string' && /^[0-9A-Za-z]{22}$/.test(args[1])) {
			args[1] = this.decode(args[1]);
		}
		const id = uuidv5.apply(this, args);
		return this.encode(id);
	}

	v1() {
		const args = Array.prototype.slice.call(arguments);
		if (!args[1]) {
			// make sure we use a buffer to avoid getting a uuid with dashes
			args[1] = new Buffer(16);
		}
		const id = uuidv1.apply(this, args);
		return this.encode(id);
	}

	encode(input, encoding) {
		encoding = encoding || 'hex';
		if (typeof input === 'string') {
			// remove the dashes to save some space
			input = new Buffer(input.replace(/-/g, ''), encoding);
		}
		return ensureLength(this.base.encode(input), OUTPUT_LENGTH);
	}

	decode(input, encoding) {
		encoding = encoding || 'hex';
		const res = ensureLength(
			new Buffer(this.base.decode(input)).toString(encoding),
			UUID_LENGTH
		);
		// insert dashes on return
		return `${res.slice(0, 8)}-${res.slice(8, 12)}-${res.slice(12, 16)}-${res.slice(16, 20)}-${res.slice(20)}`;
	}
}

function ensureLength(input, targetLength) {
	input = input.toString();
	return `${'0'.repeat(32)}${input}`.slice(-targetLength);
}

module.exports = new UUID62();
